Sources/NIOSOCKS/State/ClientStateMachine.swift

```
@@ -0,0 +1,151 @@
+                return .waitForMoreData
+            }
+            guard greeting.methods.contains(selected.method) else {
+                throw SOCKSError.InvalidAuthenticationSelection(selection: selected.method)
+            }
+                
+            // we don't current support any form of authentication
+            return self.authenticate(&buffer)
+        }
+    }
+    
+    mutating func handleServerResponse(_ buffer: inout ByteBuffer, request: ClientRequest) throws -> ClientAction {
+        return try self.unwindIfNeeded(&buffer) { buffer -> ClientAction in
+            guard let response = try buffer.readServerResponse() else {
+                return .waitForMoreData
+            }
+            guard response.reply == .succeeded else {
+                throw SOCKSError.ConnectionFailed(reply: response.reply)
+            }
+            self.state = .active
+            return .proxyEstablished
+        }
+    }
+    
+    mutating func authenticate(_ buffer: inout ByteBuffer) -> ClientAction {
+        return self.unwindIfNeeded(&buffer) { buffer -> ClientAction in
+            
+            // we don't currently support any authentication
+            // so assume all is fine, and instruct the client
+            // to send the request
+            self.state = .waitingForClientRequest
+            return .sendRequest
+        }
+    }
+    
+}
+
+// MARK: - Outgoing
+extension ClientStateMachine {
+    
+    mutating func connectionEstablished() -> ClientAction {
+        assert(self.state == .inactive)
+        self.state = .waitingForClientGreeting
+        return .sendGreeting
+    }
+
+    mutating func sendClientGreeting(_ greeting: ClientGreeting) {
+        assert(self.state == .waitingForClientGreeting)
+        self.state = .waitingForAuthenticationMethod(greeting)
+    }
+    
+    mutating func sendClientRequest(_ request: ClientRequest) {
+        assert(self.state == .waitingForClientRequest)
```

I'd rather that these actually handled the error case by throwing, instead of silently tolerating the error in production code.